### PR TITLE
Send emails to pending users

### LINKED ISSUE #1900: Email pending users when added to database
Although the interface suggests this is happening, it was actually never implemented.

screenshot from 2015-05-28 13 34 17

> Note: If this person does not already have an Indico account, 
he or she will be sent an email asking to register as a user.
After the registration the user will automatically be given
submission rights.

This should probably be triggered in `indico.modules.users.util.get_user_by_email`.

### PR DESCRIPTION
**Pull request by mic4ael**
Sending an email to newly created user when creating a contribution.

fixes #1900

### CODE REVIEW

File: indico/util/user.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 # You should have received a copy of the GNU General Public License
 # along with Indico; if not, see <http://www.gnu.org/licenses/>.
 
+import re
+
```
Comment (ThiefMaster) [inline indico/util/user.py]: no linebreak here. re and functools are both stdlib

File: indico/util/user.py [hunk 2]
```diff
@@ -153,3 +155,25 @@ def wrapper(*args, **kwargs):
...
 
     return wrapper
+
+
+def create_pending_user(user_data):
+    from indico.modules.users import User
+    from indico.modules.users.models.users import UserTitle
+
+    new_user = User()
+    new_user.email = user_data.get('email').lower()
```
Comment (ThiefMaster) [inline indico/util/user.py]: this kind of data is required. i'd rather use a separate argument for it or at least use `[]` so it fails if it's missing

File: indico/util/user.py [hunk 3]
```diff
@@ -153,3 +155,25 @@ def wrapper(*args, **kwargs):
...
+    new_user.first_name = user_data.get('firstName')
```
Comment (ThiefMaster) [inline indico/util/user.py]: this forces you to pass camelcased data...

File: indico/util/user.py [hunk 4]
```diff
@@ -153,3 +155,25 @@ def wrapper(*args, **kwargs):
...
+    new_user.last_name = user_data.get('familyName')
+    new_user.title = int(UserTitle.get(re.sub(r'[^A-Za-z]', '', user_data.get('title').lower())))
```
Comment (ThiefMaster) [inline indico/util/user.py]: that looks quite hacky... also, no need to convert to `int`.

File: indico/util/user.py [hunk 5]
```diff
@@ -153,3 +155,25 @@ def wrapper(*args, **kwargs):
...
+    new_user.is_pending = True
+
+    db.session.add(new_user)
+    db.session.flush()
+    return new_user
+
+
+def send_email_to_pending_contributors(email, event_name):
+    from indico.modules.auth.controllers import _send_confirmation
```
Comment (ThiefMaster) [inline indico/util/user.py]: this shouldn't be underscored if you use it outside the module

File: indico/util/user.py (hunk 2 above)
Comment (mic4ael) [inline indico/util/user.py]: Good remark. Thanks.

File: indico/util/user.py (hunk 5 above)
Comment (mic4ael) [inline indico/util/user.py]: Sure, however changing the name of this function could result in problems after merging one of my Pull Requests.

File: indico/util/user.py (hunk 3 above)
Comment (mic4ael) [inline indico/util/user.py]: Can't change it easily cause it can cause a lot of backward-compatibility related problems.

Comment (ThiefMaster) [inline indico/util/user.py]: `create_pending_user` is not a legacy function. If you need to integrate it with lots of legacy code, maybe add another function that calls it with proper data

Comment (mic4ael) [inline indico/util/user.py]: Like for example `unify_user_args`?

Comment (ThiefMaster) [inline indico/util/user.py]: no... just a plain function that takes the ugly dict and calls `create_pending_user` with proper arguments

or even easier, just take the values from the dict and pass them as normal arguments to `create_pending_user`..

File: indico/util/user.py [hunk 6]
```diff
@@ -153,3 +154,36 @@ def wrapper(*args, **kwargs):
...
+    def _wrapper(user_data):
+        user_data['first_name'] = user_data.pop('firstName')
+        user_data['family_name'] = user_data.pop('familyName')
+        return decorated_func(user_data)
+
+    return _wrapper
+
+
+@_unify_pending_user_data
+def create_pending_user(user_data):
```
Comment (ThiefMaster) [inline indico/util/user.py]: docstring missing

File: indico/util/user.py [hunk 7]
```diff
@@ -153,3 +154,36 @@ def wrapper(*args, **kwargs):
         return fn(*args, **kwargs)
 
     return wrapper
+
+
+def _unify_pending_user_data(decorated_func):
```
Comment (ThiefMaster) [inline indico/util/user.py]: usually the arg is called `fn`, `func` or `f` in a decorator

**Comment by pferreir** [comment]
We should make sure that we are telling the user an e-mail will be sent to the person in question. I'm not sure this is happening in all cases.

File: indico/util/user.py (hunk 6 above)
Comment (pferreir) [inline indico/util/user.py]: Can't we use that [here](https://github.com/indico/indico/blob/master/indico/modules/users/util.py#L198)?

Comment (mic4ael) [inline indico/util/user.py]: Probably we can ;)

File: indico/modules/auth/templates/emails/verify_and_register.txt [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    {% trans %}Verify your email{% endtrans %}
+{%- endblock %}
+
+{% block header_recipient -%}
+    {% trans %}user{% endtrans %}
+{%- endblock %}
+
+{% block body -%}
+    {%- filter dedent -%}
+            To have a full access to its details, please verify that e-mail address {{ email }} is yours indeed.
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/emails/verify_and_register.txt]: weird indentation

File: indico/util/user.py [hunk 9]
```diff
@@ -153,3 +154,46 @@ def wrapper(*args, **kwargs):
...
+    """Helper function to create a pending user.
+
+    :param user_data: data used to populate new User object
+
+    """
+    from indico.modules.users import User
+    from indico.modules.users.models.users import UserTitle
+
+    new_user = User()
+    new_user.first_name = user_data['first_name']
```
Comment (ThiefMaster) [inline indico/util/user.py]: all those fields should be passed through `to_unicode()`

**Comment by ThiefMaster** [comment]
Obsolete since we don't create pending users anymore.

Could be done using an acl change signal though like it's done for pending event submitters. See `_notify_pending` in the events module

**Comment by mic4ael** [comment]
In that case, what should we do right now?

**Comment by ThiefMaster** [comment]
Actually, we still create pending users when adding someone to an ACL who is not in Indico yet. However, if i remember the last time discussing this with @pferreir correctly we do not want to send emails to users automatically when adding them somewhere in an event. This would be an explicit action, probably from the Roles page.

**Comment by mic4ael** [comment]
Then I guess I can close this PR.

**Pull request closed without merging**